src/planner/select_plan.go

```diff
@@ -71,73 +82,65 @@ func NewSelectPlan(log *xlog.Log, database string, query string, node *sqlparser
 // Unsupports:
 // 1. subquery
-func (p *SelectPlan) analyze() (string, string, string, error) {
-	var shardDatabase string
-	var shardTable string
-	var aliasTable string
-	var tableExpr *sqlparser.AliasedTableExpr
+func (p *SelectPlan) analyze() ([]TableInfo, error) {
+	tableInfos := make([]TableInfo, 0, 4)
 	node := p.node
 
 	// Check subquery.
 	if hasSubquery(node) || len(node.From) > 1 {
-		return shardDatabase, shardTable, aliasTable, errors.New("unsupported: subqueries.in.select")
+		return nil, errors.New("unsupported: subqueries.in.select")
 	}
 
-	// Find the first table in the node.From.
-	// Currently only support AliasedTableExpr, JoinTableExpr select.
+	// Get table info in the node.From.
+	// Only support AliasedTableExpr, JoinTableExpr select.
 	switch expr := (node.From[0]).(type) {
 	case *sqlparser.AliasedTableExpr:
-		tableExpr = expr
-	case *sqlparser.JoinTableExpr:
-		if v, ok := (expr.LeftExpr).(*sqlparser.AliasedTableExpr); ok {
-			tableExpr = v
+		tableInfo, err := p.getOneTableInfo(expr)
+		if err != nil {
+			return nil, err
 		}
-	}
-
-	if tableExpr != nil {
-		aliasTable = tableExpr.As.String()
-
-		switch expr := tableExpr.Expr.(type) {
-		case sqlparser.TableName:
-			if !expr.Qualifier.IsEmpty() {
-				shardDatabase = expr.Qualifier.String()
-			}
-			shardTable = expr.Name.String()
+		tableInfos = append(tableInfos, tableInfo)
+	case *sqlparser.JoinTableExpr:
+		if err := p.getJoinTableInfos(expr, &tableInfos); err != nil {
+			return nil, err
 		}
+	default:
+		return nil, errors.New("unsupported: ParenTableExpr.in.select")
 	}
-	return shardDatabase, shardTable, aliasTable, nil
+	return tableInfos, nil
 }
 
 // Build used to build distributed querys.
 // For now, we don't support subquery in select.
 func (p *SelectPlan) Build() error {
-	var err error
-	var shardTable string
-	var aliasTable string
-	var shardDatabase string
-
 	log := p.log
 	node := p.node
-	if shardDatabase, shardTable, aliasTable, err = p.analyze(); err != nil {
+
+	tableInfos, err := p.analyze()
+	if err != nil {
 		return err
 	}
-	if shardDatabase == "" {
-		shardDatabase = p.database
-	}
-	if aliasTable == "" {
-		aliasTable = shardTable
+	t := tableInfos[0]
+	// Support only one shard tables
+	var num int
+	for _, tableInfo := range tableInfos {
+		if tableInfo.shardKey != "" {
+			t = tableInfo
+			num++
+		}
+		if num > 1 {
+			return errors.New("unsupported: more.than.one.shard.tables")
+		}
 	}
 
-	// Get the routing segments info.
-	shardkey, err := p.router.ShardKey(shardDatabase, shardTable)
+	segments, err := getDMLRouting(t.database, t.tableName, t.shardKey, node.Where, p.router)
 	if err != nil {
 		return err
 	}
-	segments, err := getDMLRouting(shardDatabase, shardTable, shardkey, node.Where, p.router)
-	if err != nil {
-		return err
+	// The tables are global tables, send to one backend
+	if num == 0 {
+		segments = segments[:1]
```

Comment:
Just the first segment, all load on the the first segment, Load may be Unbalanced?